dvc/repo/datasets.py

```diff
@@ -0,0 +1,350 @@
+        return cls(**kwargs)
+
+
+@frozen(kw_only=True)
+class DVCDataset(SerDe):
+    name: str
+    url: str
+    type: str
+    path: str = field(default="", converter=default_str)
+    rev: Optional[str] = None
+
+    def lock(self, repo, **kwargs):
+        from dvc.dependency import RepoDependency
+
+        def_repo = {
+            RepoDependency.PARAM_REV: self.rev,
+            RepoDependency.PARAM_URL: self.url,
+        }
+        dep = RepoDependency(def_repo, None, self.path, repo=repo)  # type: ignore[arg-type]
+        dep.save()
+        d = dep.dumpd()
+
+        repo_info = d[RepoDependency.PARAM_REPO]
+        assert isinstance(repo_info, dict)
+        rev_lock = repo_info[RepoDependency.PARAM_REV_LOCK]
+        return DVCDatasetLock(
+            name=self.name,
+            type=self.type,
+            url=self.url,
+            path=self.path,
+            rev=self.rev,
+            rev_lock=rev_lock,
+        )
+
+
+@frozen(kw_only=True)
+class DVCDatasetLock(DVCDataset):
+    rev_lock: str
+
+    def to_spec(self) -> "DVCDataset":
+        return DVCDataset.from_dict(self.to_dict())
+
+
+@frozen(kw_only=True)
+class DVCXDataset(SerDe):
+    name: str
+    url: str
+    type: str
+
+    @property
+    def name_version(self) -> tuple[str, Optional[int]]:
+        url = urlparse(self.url)
+        parts = url.netloc.split("@v")
+        assert parts
+
+        name = parts[0]
+        version = int(parts[1]) if len(parts) > 1 else None
+        return name, version
+
+    def lock(
+        self,
+        repo,  # noqa: ARG002
+        record: Optional["DatasetRecord"] = None,
+        version: Optional[int] = None,
+        **kwargs,
+    ):
+        if not record:
+            try:
+                from dvcx.catalog import get_catalog  # type: ignore[import]
+
+            except ImportError as exc:
+                raise DvcException("dvcx is not installed") from exc
+
+            name, version = self.name_version
+            catalog = get_catalog()
+            record = catalog.get_remote_dataset(name)
+        assert record is not None
+        return self._lock_from_dataset_record(record, version=version)
+
+    def _lock_from_dataset_record(
+        self, record: "DatasetRecord", version: Optional[int] = None
+    ) -> "DVCXDatasetLock":
+        ver = version or record.latest_version
+        assert ver
+        version_info = record.get_version(ver)
+        return DVCXDatasetLock(
+            name=self.name,
+            url=self.url,
+            type=self.type,
+            version=version_info.version,
+            created_at=version_info.created_at,
+        )
+
+
+@frozen(kw_only=True)
+class DVCXDatasetLock(DVCXDataset):
+    version: int
+    created_at: datetime = field(converter=to_datetime)
+
+    def to_spec(self) -> "DVCXDataset":
+        return DVCXDataset.from_dict(self.to_dict())
+
+
+@frozen(kw_only=True, order=True)
+class FileInfo(SerDe):
```

Comment:
I have split up `files` metadata into `relpath` and `meta`. In `dvc.lock`, they are a single dictionary. 

I did this to make it simpler to serialize/deserialize, and I find it conceptually better, as the `meta` is for a `path`.